rfcs/020 - Dependencies permissions RFC.md

```diff
@@ -0,0 +1,104 @@
+# 020 - Dependencies permissions RFC
+
+## Current Status
+
+### Proposed
+
+2021-08-05
+
+### Accepted
+
+YYYY-MM-DD
+
+#### Approvers
+
+- David Ellis <david@alantechnologies.com>
+- Luis De Pombo <luis@alantechnologies.com>
+- Colton Donnelly <colton@alantechnologies.com>
+
+### Implementation
+
+- [ ] Implemented: [One or more PRs](https://github.com/alantech/alan/some-pr-link-here) YYYY-MM-DD
+- [ ] Revoked/Superceded by: [RFC ###](./000 - RFC Template.md) YYYY-MM-DD
+
+## Author(s)
+
+- Alejandro Guillen <alejandro@alantechnologies.com>
+
+## Summary
+
+Alan's third-party module permission system is one of the value propositions of the language. It adds a layer of security no other project approaches. The idea is to allow users to prevent specific third-party dependencies from having access to specific standard libraries that they should not have access to. This can be achieve with current mocking built-in and some updates to the `@std/deps` standard library.
+
+## Expected SemBDD Impact
+
+This would be a minor update if we were post-1.0 as it should have zero breaking impact on existing code.
+
+## Proposal
+
+The idea is to take advantage of alan's module resolution and build-in mock systems to prevent libraries to access specific standard libraries. We could also block specific standard libraries at application level, meaning that we would not want that any of our third party libraries use them. If we do not specify anything and just use the library will use everything.
+
+The priorities of the module resoulution system would go as follow starting with the highest priority:
+
+1. Actual dependency level `modules` directory
+2. Global dependencies `modules` directory
+3. Application `modules` directory
+
+Meaning that inner `modules` > outer `modules`.
+
+### Block at library level
+
+This is the basic case when we want to use `new_dep_A` in our project but we do not want this library to have access to `std/cmd` for example. So, we we add this library we will specify the standard libraries blacklisted. This block would need to work recursively due to module resolution priorities. The flow will be:
+
+![Block at library level](./lib-lvl-v1.png)
+
+The `.dependencies.ln` file could look something like:
+
+Block `@std/cmd` but do not override if any mock exists:
+```ln
+add('https://github.com/org/new_dep_A', some(['cmd']), false);
```

Comment:
First comment: That should probably be `some(['@std/cmd'])`. It is possible that someone would want to restrict access to third-party dependencies that the library has declared a dependency on.

Second comment: This is an awkward API and the meaning of the arguments aren't clear while reading the code itself.

Some ideas on how to make this a nicer API:

### 1. Utilize multiple functions with the same name

```js
add('https://github.com/org/new_dep_A', ['@std/cmd']); // Your first example
add('https://github.com/org/new_dep_A', ['@std/cmd'], true); // Your second example
add('https://github.com/org/new_dep_A'); // Your third example
```

You can define `fn add(url: string), fn add(url: string, blockedLibs: Array<string>), fn add(url: string, blockedLibs: Array<string>, fullBlock: bool)` and it'll select the right one in the right situation.

This still doesn't explain *what* the args mean, though.

### 2. Make it a fluent API

```js
'https://github.com/org/new_dep_A'.block('@std/cmd').add(); // Your first example
'https://github.com/org/new_dep_A'.block('@std/cmd').block('@std/tcp').fullOverride().add(); // Your second example
'https://github.com/org/new_dep_A'.block(['@std/cmd', '@std/tcp']).fullOverride().add(); // Another form of your second example
'https://github.com/org/new_dep_A'.add() // Your third example
```

But even this has some issues (what if you want to override a mock for `@std/cmd` but not for `@std/tcp`?) So maybe this would be better:

```js
'https://github.com/org/new_dep_A'.block('@std/cmd').add(); // Your first example
'https://github.com/org/new_dep_A'.block('@std/cmd').fullBlock('@std/tcp').add(); // Your second example
'https://github.com/org/new_dep_A'.block(['@std/cmd', '@std/tcp']).add(); // Another form of your second example
'https://github.com/org/new_dep_A'.add() // Your third example
```

In either case, the implementation is something like this: `fn add(url: string)` for the basic situation, but for the rest there's a `Dependency` struct that tracks the metadata that you are building and the `fn add(dep: Dependency)` function parses that struct into the correct set of rules to build. `fn add(url: string)` could simply create one of these `Dependency` values and then call the other add on it to keep things DRY.

### 3. Expose that `Dependency` type directly

Something like:

```js
add(new Dependency {
  url: 'https://github.com/org/new_dep_A',
  block: ['@std/cmd'],
  fullBlock: [],
}); // First example

new Dependency {
  url: 'https://github.com/org/new_dep_A',
  block: [],
  fullBlock: ['@std/cmd'],
}.add(); // Second example

add(new Dependency {
  url: 'https://github.com/org/new_dep_A',
  block: [],
  fullBlock: [],
}); // Third example
```

This makes it very clear what's going on all the time, but requires the author to be aware of all of the installation options and to provide a value for each of them, which may or may not be a bad thing?

Finally, these three options are all compatible with each other (a multi-arg `add` could construct a `Dependency` value internally). I would personally only go with a combo of 2 and 3, though.

Reply:
On your first comment, yes, I had in mind only standard library blocks but yes there's no reason to limit it only to that.

On your second comment, completely agree on your suggestion, the API I present was quite confusing. It was a naive approach to express the flow and understand all the pieces.

I'll think about it but now I like the option of a `fn add(url: string): Dependency` and then calling `block` of `fullBlock` (not so sure about this name) over `Dependency`.

Reply:
> So maybe this would be better:
> 
> ```js
> 'https://github.com/org/new_dep_A'.block('@std/cmd').add(); // Your first example
> 'https://github.com/org/new_dep_A'.block('@std/cmd').fullBlock('@std/tcp').add(); // Your second example
> 'https://github.com/org/new_dep_A'.block(['@std/cmd', '@std/tcp']).add(); // Another form of your second example
> 'https://github.com/org/new_dep_A'.add() // Your third example
> ```
> 

I personally think this is the best solution. The functions in `@std/deps` would look like:
```ts
fn add(url: string) = new Dependency {
  url: url,
  block: [],
  fullBlock: [],
}.add();
fn add(dep: Dependency) = adddep(dep);
fn block(dep: Dependency, block: string): Dependency {
  dep.block.push(block);
  return dep;
}
fn block(dep: Dependency, block: Array<string>): Dependency {
  dep.block = dep.block + block;
  return dep;
}
// repeat both `block` fns above but `s/block/fullBlock`
```
(btw, realizing that I don't believe I have proper detection on fields vs method calls in lnn right now, I'll have to fix that)

> 
> ### 3. Expose that `Dependency` type directly
> Something like:
> 
> ```js
> add(new Dependency {
>   url: 'https://github.com/org/new_dep_A',
>   block: ['@std/cmd'],
>   fullBlock: [],
> }); // First example
> 
> new Dependency {
>   url: 'https://github.com/org/new_dep_A',
>   block: [],
>   fullBlock: ['@std/cmd'],
> }.add(); // Second example
> 
> add(new Dependency {
>   url: 'https://github.com/org/new_dep_A',
>   block: [],
>   fullBlock: [],
> }); // Third example
> ```
> 
> This makes it very clear what's going on all the time, but requires the author to be aware of all of the installation options and to provide a value for each of them, which may or may not be a bad thing?

Most other dependency configuration tools don't require all fields to be populated, since the focus is on how easy it is to add dependencies. If users need more configuration options, then it's usually a minor change to add those configurations. For example, `Cargo.toml` values in the `dependencies` table can either be strings (indicating a semver-compatible string) or an object with required key `version` and optional keys `features`, `no-default-features`, `optional`, etc. A similar mechanism is present in npm, gemspec, and pubspec - some other popular package managers I looked at (cabal, stack, pip, opam) only support specifying the version (pip is weird and can do more, but not via table-types)